File: typescript/05_loops.ts
// ============================================================
// Topic:  Loops
// Summary: C-style for, for..of, for..in, while, entries, array methods.
// Run:     npx ts-node typescript/05_loops.ts
// ============================================================

// ----- 1. Classic C-style for (Python uses range() instead) -----
for (let i = 0; i < 3; i++) {
  process.stdout.write(i + " "); // Output: 0 1 2
}
console.log();

// ----- 2. for..of -> values (Python: for x in coll) -----
const fruits = ["apple", "banana", "cherry"];
for (const fruit of fruits) {
  process.stdout.write(fruit + " "); // Output: apple banana cherry
}
console.log();

// ----- 3. entries() -> index + value (Python: enumerate) -----
for (const [index, fruit] of fruits.entries()) {
  process.stdout.write(`${index}:${fruit} `); // Output: 0:apple 1:banana 2:cherry
}
console.log();

// ----- 4. for..in -> KEYS/indexes (⚠ not values! different from Python) -----
for (const key in fruits) {
  process.stdout.write(key + " "); // Output: 0 1 2  (indexes as strings)
}
console.log();

// ----- 5. while -----
let n = 3;
while (n > 0) {
  process.stdout.write(n + " "); // Output: 3 2 1
  n--;
}
console.log();

// ----- 6. break and continue -----
for (let i = 0; i < 10; i++) {
  if (i === 3) continue; // skip 3
  if (i === 6) break; // stop at 6
  process.stdout.write(i + " "); // Output: 0 1 2 4 5
}
console.log();

// ----- 7. Functional iteration (Python: list comprehension) -----
const squares = [0, 1, 2, 3, 4].map((x) => x * x);
console.log(squares); // Output: [ 0, 1, 4, 9, 16 ]
const evens = [...Array(10).keys()].filter((x) => x % 2 === 0);
console.log(evens); // Output: [ 0, 2, 4, 6, 8 ]

// Cross-language note:
//   TS `for..of` == Python `for x in coll` (values).
//   TS `for..in` gives KEYS — NOT the same as Python's `for x in` (values). Careful!
//   TS replaces range() with the C-style counter loop or [...Array(n).keys()].

File: java/05_loops.java
// ============================================================
// Topic:  Loops
// Summary: C-style for, for-each, while, break/continue, streams.
// Run:     java java/05_loops.java   (Java 11+ single-file launch)
// ============================================================

import java.util.ArrayList;
import java.util.List;

public class Loops {
    public static void main(String[] args) {
        // ----- 1. C-style for (counter loop) -----
        for (int i = 0; i < 3; i++) {
            System.out.print(i + " ");       // Output: 0 1 2
        }
        System.out.println();

        // range with step (Python: range(2, 11, 2))
        for (int i = 2; i < 11; i += 2) {
            System.out.print(i + " ");       // Output: 2 4 6 8 10
        }
        System.out.println();

        // ----- 2. Enhanced for / for-each (Python: for x in coll; TS: for..of) -----
        List<String> fruits = new ArrayList<>(List.of("apple", "banana", "cherry"));
        for (String fruit : fruits) {
            System.out.print(fruit + " ");   // Output: apple banana cherry
        }
        System.out.println();

        // ----- 3. index + value (Python: enumerate) -----
        for (int i = 0; i < fruits.size(); i++) {
            System.out.print(i + ":" + fruits.get(i) + " "); // Output: 0:apple 1:banana 2:cherry
        }
        System.out.println();

        // ----- 4. while loop -----
        int n = 3;
        while (n > 0) {
            System.out.print(n + " ");
            n--;
        }
        System.out.println();                // Output: 3 2 1

        // ----- 5. break and continue -----
        for (int i = 0; i < 10; i++) {
            if (i == 3) continue;            // skip 3
            if (i == 6) break;               // stop at 6
            System.out.print(i + " ");       // Output: 0 1 2 4 5
        }
        System.out.println();

        // ----- 6. Build a list (Python: list comprehension; TS: map) -----
        List<Integer> squares = new ArrayList<>();
        for (int i = 0; i < 5; i++) squares.add(i * i);
        System.out.println(squares);         // Output: [0, 1, 4, 9, 16]

        // Cross-language note:
        //   Enhanced for `for (T x : coll)` == Python `for x in coll` == TS `for..of`.
        //   No range() built-in — use a C-style for or IntStream.range(0, n).
        //   Java has i++; Python uses i += 1.
    }
}
